#Written by Malachi Griffith
#Copyright 2009 Malachi Griffith
#This file is part of 'ALEXA-Seq'
#ALEXA-Seq is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
#ALEXA-Seq is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#You should have received a copy of the GNU General Public License along with ALEXA-Seq (COPYING.txt).  If not, see <http://www.gnu.org/licenses/>.

#Compare Gene expression estimates generated by each platform
#Affmetrix Exon arrays, ALEXA arrays, Solexa WTSS
#All based on gene models from the same version of EnsEMBL 

library(geneplotter)
library(RColorBrewer)
library(gcrma)

datadir = "C:/Documents and Settings/MG/Desktop/5-FU Resistance WTSS-Array Paper/Illumina Sequence Analysis/Expression and Alternative Expression/Cross Platform Comparisons/gene"
setwd(datadir)
dir()

affy_wtss_data_p1 = read.table("AFFY_Plus1_AND_WTSS_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
affy_wtss_data_p4 = read.table("AFFY_Plus4_AND_WTSS_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
affy_wtss_data_p16 = read.table("AFFY_Plus16_AND_WTSS_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
alexa_wtss_data_p1 = read.table("ALEXA_Plus1_AND_WTSS_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
alexa_wtss_data_p4 = read.table("ALEXA_Plus4_AND_WTSS_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
alexa_wtss_data_p16 = read.table("ALEXA_Plus16_AND_WTSS_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
alexa_affy_data_p1 = read.table("ALEXA_AND_AFFY_Plus1_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
alexa_affy_data_p4 = read.table("ALEXA_AND_AFFY_Plus4_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
alexa_affy_data_p16 = read.table("ALEXA_AND_AFFY_Plus16_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")

wtss_data_all_genes_p1 = read.table("WTSS_Data_ALL_Genes_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
wtss_data_all_genes_p4 = read.table("WTSS_Data_ALL_Genes_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")
wtss_data_all_genes_p16 = read.table("WTSS_Data_ALL_Genes_EnsEMBL49.txt", header = TRUE, sep = "\t", comment.char = "#")

############################################################################################################################
#1.) First determine which variance stabalization adjustment value gives the best cross platform comparisons (1, 4 or 16)
#    ANSWER - doesnt seem to make a big difference
#    NOTE: limiting comparison to only those ~10,000 genes expressed by WTSS at 10X NORM1 coverage or higher, greatly improves correlation
#    At some point should replace 10X value with one that is chosen less arbitrarily
#    Also determine which style of expression values gives the best correlations to array data:
#    A.) Read counts. (Actual read count) 
#    B.) Average Coverage - RAW (Cumulative base coverage divided by gene length) 
#    C.) Average Coverage - NORM1 values. (RAW values scaled to smallest library)
#    D.) Average Coverage - NORM2 values. (RAW values scaled to smallest library AND corrected for mapability)
#    E.) Average Coverage - NORM3 values. (RAW values quantiles normalized)

# Note that DE values calculated from (A) and (B) should be identical...

adjust_1 = 1
adjust_4 = 4
adjust_16 = 16

#A.) Calculate DE values for all WTSS data points - using 'Read Count' WTSS values (simply raw read counts)
affy_wtss_data_p1[,"wtss_log2_de_rc"] = log2(affy_wtss_data_p1[,"MIP101_Read_Count"] + adjust_1) - log2(affy_wtss_data_p1[,"MIP5FUR_Read_Count"] + adjust_1)
affy_wtss_data_p4[,"wtss_log2_de_rc"] = log2(affy_wtss_data_p4[,"MIP101_Read_Count"] + adjust_4) - log2(affy_wtss_data_p4[,"MIP5FUR_Read_Count"] + adjust_4)
affy_wtss_data_p16[,"wtss_log2_de_rc"] = log2(affy_wtss_data_p16[,"MIP101_Read_Count"] + adjust_16) - log2(affy_wtss_data_p16[,"MIP5FUR_Read_Count"] + adjust_16)

alexa_wtss_data_p1[,"wtss_log2_de_rc"] = log2(alexa_wtss_data_p1[,"MIP101_Read_Count"] + adjust_1) - log2(alexa_wtss_data_p1[,"MIP5FUR_Read_Count"] + adjust_1)
alexa_wtss_data_p4[,"wtss_log2_de_rc"] = log2(alexa_wtss_data_p4[,"MIP101_Read_Count"] + adjust_4) - log2(alexa_wtss_data_p4[,"MIP5FUR_Read_Count"] + adjust_4)
alexa_wtss_data_p16[,"wtss_log2_de_rc"] = log2(alexa_wtss_data_p16[,"MIP101_Read_Count"] + adjust_16) - log2(alexa_wtss_data_p16[,"MIP5FUR_Read_Count"] + adjust_16)

wtss_data_all_genes_p1[,"wtss_log2_de_rc"] = log2(wtss_data_all_genes_p1[,"MIP101_Read_Count"] + adjust_1) - log2(wtss_data_all_genes_p1[,"MIP5FUR_Read_Count"] + adjust_1)
wtss_data_all_genes_p4[,"wtss_log2_de_rc"] = log2(wtss_data_all_genes_p4[,"MIP101_Read_Count"] + adjust_4) - log2(wtss_data_all_genes_p4[,"MIP5FUR_Read_Count"] + adjust_4)
wtss_data_all_genes_p16[,"wtss_log2_de_rc"] = log2(wtss_data_all_genes_p16[,"MIP101_Read_Count"] + adjust_16) - log2(wtss_data_all_genes_p16[,"MIP5FUR_Read_Count"] + adjust_16)

#B.) Calculate DE values for all WTSS data points - using 'Average Coverage - RAW' WTSS values
affy_wtss_data_p1[,"wtss_log2_de_raw"] = log2(affy_wtss_data_p1[,"MIP101_Average_Coverage_RAW"] + adjust_1) - log2(affy_wtss_data_p1[,"MIP5FUR_Average_Coverage_RAW"] + adjust_1)
affy_wtss_data_p4[,"wtss_log2_de_raw"] = log2(affy_wtss_data_p4[,"MIP101_Average_Coverage_RAW"] + adjust_4) - log2(affy_wtss_data_p4[,"MIP5FUR_Average_Coverage_RAW"] + adjust_4)
affy_wtss_data_p16[,"wtss_log2_de_raw"] = log2(affy_wtss_data_p16[,"MIP101_Average_Coverage_RAW"] + adjust_16) - log2(affy_wtss_data_p16[,"MIP5FUR_Average_Coverage_RAW"] + adjust_16)

alexa_wtss_data_p1[,"wtss_log2_de_raw"] = log2(alexa_wtss_data_p1[,"MIP101_Average_Coverage_RAW"] + adjust_1) - log2(alexa_wtss_data_p1[,"MIP5FUR_Average_Coverage_RAW"] + adjust_1)
alexa_wtss_data_p4[,"wtss_log2_de_raw"] = log2(alexa_wtss_data_p4[,"MIP101_Average_Coverage_RAW"] + adjust_4) - log2(alexa_wtss_data_p4[,"MIP5FUR_Average_Coverage_RAW"] + adjust_4)
alexa_wtss_data_p16[,"wtss_log2_de_raw"] = log2(alexa_wtss_data_p16[,"MIP101_Average_Coverage_RAW"] + adjust_16) - log2(alexa_wtss_data_p16[,"MIP5FUR_Average_Coverage_RAW"] + adjust_16)

wtss_data_all_genes_p1[,"wtss_log2_de_raw"] = log2(wtss_data_all_genes_p1[,"MIP101_Average_Coverage_RAW"] + adjust_1) - log2(wtss_data_all_genes_p1[,"MIP5FUR_Average_Coverage_RAW"] + adjust_1)
wtss_data_all_genes_p4[,"wtss_log2_de_raw"] = log2(wtss_data_all_genes_p4[,"MIP101_Average_Coverage_RAW"] + adjust_4) - log2(wtss_data_all_genes_p4[,"MIP5FUR_Average_Coverage_RAW"] + adjust_4)
wtss_data_all_genes_p16[,"wtss_log2_de_raw"] = log2(wtss_data_all_genes_p16[,"MIP101_Average_Coverage_RAW"] + adjust_16) - log2(wtss_data_all_genes_p16[,"MIP5FUR_Average_Coverage_RAW"] + adjust_16)

#C.) Calculate DE values for all WTSS data points - using 'NORM1' WTSS values (i.e. scaled to smaller library)
affy_wtss_data_p1[,"wtss_log2_de_n1"] = log2(affy_wtss_data_p1[,"MIP101_Average_Coverage_NORM1"] + adjust_1) - log2(affy_wtss_data_p1[,"MIP5FUR_Average_Coverage_NORM1"] + adjust_1)
affy_wtss_data_p4[,"wtss_log2_de_n1"] = log2(affy_wtss_data_p4[,"MIP101_Average_Coverage_NORM1"] + adjust_4) - log2(affy_wtss_data_p4[,"MIP5FUR_Average_Coverage_NORM1"] + adjust_4)
affy_wtss_data_p16[,"wtss_log2_de_n1"] = log2(affy_wtss_data_p16[,"MIP101_Average_Coverage_NORM1"] + adjust_16) - log2(affy_wtss_data_p16[,"MIP5FUR_Average_Coverage_NORM1"] + adjust_16)

alexa_wtss_data_p1[,"wtss_log2_de_n1"] = log2(alexa_wtss_data_p1[,"MIP101_Average_Coverage_NORM1"] + adjust_1) - log2(alexa_wtss_data_p1[,"MIP5FUR_Average_Coverage_NORM1"] + adjust_1)
alexa_wtss_data_p4[,"wtss_log2_de_n1"] = log2(alexa_wtss_data_p4[,"MIP101_Average_Coverage_NORM1"] + adjust_4) - log2(alexa_wtss_data_p4[,"MIP5FUR_Average_Coverage_NORM1"] + adjust_4)
alexa_wtss_data_p16[,"wtss_log2_de_n1"] = log2(alexa_wtss_data_p16[,"MIP101_Average_Coverage_NORM1"] + adjust_16) - log2(alexa_wtss_data_p16[,"MIP5FUR_Average_Coverage_NORM1"] + adjust_16)

wtss_data_all_genes_p1[,"wtss_log2_de_n1"] = log2(wtss_data_all_genes_p1[,"MIP101_Average_Coverage_NORM1"] + adjust_1) - log2(wtss_data_all_genes_p1[,"MIP5FUR_Average_Coverage_NORM1"] + adjust_1)
wtss_data_all_genes_p4[,"wtss_log2_de_n1"] = log2(wtss_data_all_genes_p4[,"MIP101_Average_Coverage_NORM1"] + adjust_4) - log2(wtss_data_all_genes_p4[,"MIP5FUR_Average_Coverage_NORM1"] + adjust_4)
wtss_data_all_genes_p16[,"wtss_log2_de_n1"] = log2(wtss_data_all_genes_p16[,"MIP101_Average_Coverage_NORM1"] + adjust_16) - log2(wtss_data_all_genes_p16[,"MIP5FUR_Average_Coverage_NORM1"] + adjust_16)

#D.) Calculate DE values for all WTSS data points - using 'Norm2' WTSS values (scaled and corrected for mapability)
affy_wtss_data_p1[,"wtss_log2_de_n2"] = log2(affy_wtss_data_p1[,"MIP101_Average_Coverage_NORM2"] + adjust_1) - log2(affy_wtss_data_p1[,"MIP5FUR_Average_Coverage_NORM2"] + adjust_1)
affy_wtss_data_p4[,"wtss_log2_de_n2"] = log2(affy_wtss_data_p4[,"MIP101_Average_Coverage_NORM2"] + adjust_4) - log2(affy_wtss_data_p4[,"MIP5FUR_Average_Coverage_NORM2"] + adjust_4)
affy_wtss_data_p16[,"wtss_log2_de_n2"] = log2(affy_wtss_data_p16[,"MIP101_Average_Coverage_NORM2"] + adjust_16) - log2(affy_wtss_data_p16[,"MIP5FUR_Average_Coverage_NORM2"] + adjust_16)

alexa_wtss_data_p1[,"wtss_log2_de_n2"] = log2(alexa_wtss_data_p1[,"MIP101_Average_Coverage_NORM2"] + adjust_1) - log2(alexa_wtss_data_p1[,"MIP5FUR_Average_Coverage_NORM2"] + adjust_1)
alexa_wtss_data_p4[,"wtss_log2_de_n2"] = log2(alexa_wtss_data_p4[,"MIP101_Average_Coverage_NORM2"] + adjust_4) - log2(alexa_wtss_data_p4[,"MIP5FUR_Average_Coverage_NORM2"] + adjust_4)
alexa_wtss_data_p16[,"wtss_log2_de_n2"] = log2(alexa_wtss_data_p16[,"MIP101_Average_Coverage_NORM2"] + adjust_16) - log2(alexa_wtss_data_p16[,"MIP5FUR_Average_Coverage_NORM2"] + adjust_16)

wtss_data_all_genes_p1[,"wtss_log2_de_n2"] = log2(wtss_data_all_genes_p1[,"MIP101_Average_Coverage_NORM2"] + adjust_1) - log2(wtss_data_all_genes_p1[,"MIP5FUR_Average_Coverage_NORM2"] + adjust_1)
wtss_data_all_genes_p4[,"wtss_log2_de_n2"] = log2(wtss_data_all_genes_p4[,"MIP101_Average_Coverage_NORM2"] + adjust_4) - log2(wtss_data_all_genes_p4[,"MIP5FUR_Average_Coverage_NORM2"] + adjust_4)
wtss_data_all_genes_p16[,"wtss_log2_de_n2"] = log2(wtss_data_all_genes_p16[,"MIP101_Average_Coverage_NORM2"] + adjust_16) - log2(wtss_data_all_genes_p16[,"MIP5FUR_Average_Coverage_NORM2"] + adjust_16)

#E.) Calculate DE values for all WTSS data points - using QUANTILES Normalized WTSS values (rather than simple scaling)
data_simple_norm = as.data.frame(normalize.quantiles(as.matrix(data.frame(affy_wtss_data_p1[,"MIP101_Average_Coverage_RAW"], affy_wtss_data_p1[,"MIP5FUR_Average_Coverage_RAW"]))))
affy_wtss_data_p1[,"wtss_log2_de_n3"] = log2(data_simple_norm[,1] + adjust_1) - log2(data_simple_norm[,2] + adjust_1)
data_simple_norm = as.data.frame(normalize.quantiles(as.matrix(data.frame(affy_wtss_data_p4[,"MIP101_Average_Coverage_RAW"], affy_wtss_data_p4[,"MIP5FUR_Average_Coverage_RAW"]))))
affy_wtss_data_p4[,"wtss_log2_de_n3"] = log2(data_simple_norm[,1] + adjust_1) - log2(data_simple_norm[,2] + adjust_1)
data_simple_norm = as.data.frame(normalize.quantiles(as.matrix(data.frame(affy_wtss_data_p16[,"MIP101_Average_Coverage_RAW"], affy_wtss_data_p16[,"MIP5FUR_Average_Coverage_RAW"]))))
affy_wtss_data_p16[,"wtss_log2_de_n3"] = log2(data_simple_norm[,1] + adjust_1) - log2(data_simple_norm[,2] + adjust_1)

data_simple_norm = as.data.frame(normalize.quantiles(as.matrix(data.frame(alexa_wtss_data_p1[,"MIP101_Average_Coverage_RAW"], alexa_wtss_data_p1[,"MIP5FUR_Average_Coverage_RAW"]))))
alexa_wtss_data_p1[,"wtss_log2_de_n3"] = log2(data_simple_norm[,1] + adjust_1) - log2(data_simple_norm[,2] + adjust_1)
data_simple_norm = as.data.frame(normalize.quantiles(as.matrix(data.frame(alexa_wtss_data_p4[,"MIP101_Average_Coverage_RAW"], alexa_wtss_data_p4[,"MIP5FUR_Average_Coverage_RAW"]))))
alexa_wtss_data_p4[,"wtss_log2_de_n3"] = log2(data_simple_norm[,1] + adjust_1) - log2(data_simple_norm[,2] + adjust_1)
data_simple_norm = as.data.frame(normalize.quantiles(as.matrix(data.frame(alexa_wtss_data_p16[,"MIP101_Average_Coverage_RAW"], alexa_wtss_data_p16[,"MIP5FUR_Average_Coverage_RAW"]))))
alexa_wtss_data_p16[,"wtss_log2_de_n3"] = log2(data_simple_norm[,1] + adjust_1) - log2(data_simple_norm[,2] + adjust_1)

data_simple_norm = as.data.frame(normalize.quantiles(as.matrix(data.frame(wtss_data_all_genes_p1[,"MIP101_Average_Coverage_RAW"], wtss_data_all_genes_p1[,"MIP5FUR_Average_Coverage_RAW"]))))
wtss_data_all_genes_p1[,"wtss_log2_de_n3"] = log2(data_simple_norm[,1] + adjust_1) - log2(data_simple_norm[,2] + adjust_1)
data_simple_norm = as.data.frame(normalize.quantiles(as.matrix(data.frame(wtss_data_all_genes_p4[,"MIP101_Average_Coverage_RAW"], wtss_data_all_genes_p4[,"MIP5FUR_Average_Coverage_RAW"]))))
wtss_data_all_genes_p4[,"wtss_log2_de_n3"] = log2(data_simple_norm[,1] + adjust_1) - log2(data_simple_norm[,2] + adjust_1)
data_simple_norm = as.data.frame(normalize.quantiles(as.matrix(data.frame(wtss_data_all_genes_p16[,"MIP101_Average_Coverage_RAW"], wtss_data_all_genes_p16[,"MIP5FUR_Average_Coverage_RAW"]))))
wtss_data_all_genes_p16[,"wtss_log2_de_n3"] = log2(data_simple_norm[,1] + adjust_1) - log2(data_simple_norm[,2] + adjust_1)


#Summarize data distributions
boxplot(wtss_data_all_genes_p1[,"wtss_log2_de_rc"], affy_wtss_data_p1[,"mean_plier_log2_DE"], alexa_wtss_data_p1[,"mean_alexa_log2_DE"],
        wtss_data_all_genes_p4[,"wtss_log2_de_rc"], affy_wtss_data_p4[,"mean_plier_log2_DE"], alexa_wtss_data_p4[,"mean_alexa_log2_DE"],
	   wtss_data_all_genes_p16[,"wtss_log2_de_rc"], affy_wtss_data_p16[,"mean_plier_log2_DE"], alexa_wtss_data_p16[,"mean_alexa_log2_DE"],
	   names=c("wt_1","af_l","al_1","wt_4","af_4","al_4","wt_16","af_l6","al_16"), main="Gene DE - WTSS Read Count (wt), Affy (af), ALEXA (al)",
	   col=c("red","blue","green","red","blue","green","red","blue","green"))

boxplot(wtss_data_all_genes_p1[,"wtss_log2_de_raw"], affy_wtss_data_p1[,"mean_plier_log2_DE"], alexa_wtss_data_p1[,"mean_alexa_log2_DE"],
        wtss_data_all_genes_p4[,"wtss_log2_de_raw"], affy_wtss_data_p4[,"mean_plier_log2_DE"], alexa_wtss_data_p4[,"mean_alexa_log2_DE"],
	   wtss_data_all_genes_p16[,"wtss_log2_de_raw"], affy_wtss_data_p16[,"mean_plier_log2_DE"], alexa_wtss_data_p16[,"mean_alexa_log2_DE"],
	   names=c("wt_1","af_l","al_1","wt_4","af_4","al_4","wt_16","af_l6","al_16"), main="Gene DE - WTSS Coverage RAW (wt), Affy (af), ALEXA (al)",
	   col=c("red","blue","green","red","blue","green","red","blue","green"))

boxplot(wtss_data_all_genes_p1[,"wtss_log2_de_n1"], affy_wtss_data_p1[,"mean_plier_log2_DE"], alexa_wtss_data_p1[,"mean_alexa_log2_DE"],
        wtss_data_all_genes_p4[,"wtss_log2_de_n1"], affy_wtss_data_p4[,"mean_plier_log2_DE"], alexa_wtss_data_p4[,"mean_alexa_log2_DE"],
	   wtss_data_all_genes_p16[,"wtss_log2_de_n1"], affy_wtss_data_p16[,"mean_plier_log2_DE"], alexa_wtss_data_p16[,"mean_alexa_log2_DE"],
	   names=c("wt_1","af_l","al_1","wt_4","af_4","al_4","wt_16","af_l6","al_16"), main="Gene DE - WTSS Coverage NORM1 (wt), Affy (af), ALEXA (al)",
	   col=c("red","blue","green","red","blue","green","red","blue","green"))

boxplot(wtss_data_all_genes_p1[,"wtss_log2_de_n2"], affy_wtss_data_p1[,"mean_plier_log2_DE"], alexa_wtss_data_p1[,"mean_alexa_log2_DE"],
        wtss_data_all_genes_p4[,"wtss_log2_de_n2"], affy_wtss_data_p4[,"mean_plier_log2_DE"], alexa_wtss_data_p4[,"mean_alexa_log2_DE"],
	   wtss_data_all_genes_p16[,"wtss_log2_de_n2"], affy_wtss_data_p16[,"mean_plier_log2_DE"], alexa_wtss_data_p16[,"mean_alexa_log2_DE"],
	   names=c("wt_1","af_l","al_1","wt_4","af_4","al_4","wt_16","af_l6","al_16"), main="Gene DE - WTSS Coverage NORM2 (wt), Affy (af), ALEXA (al)",
	   col=c("red","blue","green","red","blue","green","red","blue","green"))

boxplot(wtss_data_all_genes_p1[,"wtss_log2_de_n3"], affy_wtss_data_p1[,"mean_plier_log2_DE"], alexa_wtss_data_p1[,"mean_alexa_log2_DE"],
        wtss_data_all_genes_p4[,"wtss_log2_de_n3"], affy_wtss_data_p4[,"mean_plier_log2_DE"], alexa_wtss_data_p4[,"mean_alexa_log2_DE"],
	   wtss_data_all_genes_p16[,"wtss_log2_de_n3"], affy_wtss_data_p16[,"mean_plier_log2_DE"], alexa_wtss_data_p16[,"mean_alexa_log2_DE"],
	   names=c("wt_1","af_l","al_1","wt_4","af_4","al_4","wt_16","af_l6","al_16"), main="Gene DE - WTSS Coverage NORM3 (wt), Affy (af), ALEXA (al)",
	   col=c("red","blue","green","red","blue","green","red","blue","green"))


#Function to create correlation plot for DE values
test_corr = function(data, x_name, y_name, x_label, y_label){
  x = data[, x_name]
  y = data[, y_name]

  result1 = cor(cbind(x, y), method="pearson")
  result2 = cor(cbind(x, y), method="spearman")

  gene_count = length(x)

  result = data.frame (result1[1,2], result2[1,2], gene_count) 
  names(result) = c("pearson_all", "spearman_all", "gene_count")

  new_x_label = paste(x_label, " (", result["gene_count"], " genes)")
  main_label = paste("Gene DE,", x_label, "vs.", y_label)

  colors = colorRampPalette(c("white", "blue", "#007FFF", "cyan","#7FFF7F", "yellow", "#FF7F00", "red", "#7F0000"))

  smoothScatter(x, y, xlab=new_x_label, ylab=y_label, main=main_label, colramp=colors, nbin=275)
   abline(0,1, col="black", lwd=2)

  #Add some text to display Pearson and Spearman correlation values
  x_pos1 = max(x) * 0.60 
  y_pos1 = min(y) * 0.55
  y_pos2 = min(y) * 0.90

  text1 = paste("Pearson = ", round(result["pearson_all"], 2))
  text2 = paste("Spearman = ", round(result["spearman_all"], 2))

  text (x_pos1, y_pos1, text1)
  text (x_pos1, y_pos2, text2)

  return(result)
}

#What is the effect of using different variance stabilization values??
# x = WTSS, y = AFFY
par(mfrow=c(2,2))
test_corr(affy_wtss_data_p1, "wtss_log2_de_n1", "mean_plier_log2_DE", "WTSS+1", "AFFY+1")
test_corr(affy_wtss_data_p4, "wtss_log2_de_n1", "mean_plier_log2_DE", "WTSS+4", "AFFY+4")
test_corr(affy_wtss_data_p16, "wtss_log2_de_n1", "mean_plier_log2_DE", "WTSS+16", "AFFY+16")

# x = WTSS, y = ALEXA
par(mfrow=c(2,2))
test_corr(alexa_wtss_data_p1, "wtss_log2_de_n1", "mean_alexa_log2_DE", "WTSS+1", "ALEXA+1")
test_corr(alexa_wtss_data_p4, "wtss_log2_de_n1", "mean_alexa_log2_DE", "WTSS+4", "ALEXA+4")
test_corr(alexa_wtss_data_p16, "wtss_log2_de_n1", "mean_alexa_log2_DE", "WTSS+16", "ALEXA+16")

# x = ALEXA, y = AFFY
par(mfrow=c(2,2))
test_corr(alexa_affy_data_p1, "mip_v_fur_DE_U", "mean_plier_log2_DE", "ALEXA+1", "AFFY+1")
test_corr(alexa_affy_data_p4, "mip_v_fur_DE_U", "mean_plier_log2_DE", "ALEXA+4", "AFFY+4")
test_corr(alexa_affy_data_p16, "mip_v_fur_DE_U", "mean_plier_log2_DE", "ALEXA+16", "AFFY+16")


############################################################################################################################
#Produce three-way comparisons of DE using a variance stabilization factor of 1
# 3-way, plus1 values only
par(mfrow=c(2,2))
test_corr(affy_wtss_data_p1, "wtss_log2_de", "mean_plier_log2_DE", "WTSS", "AFFY")
test_corr(alexa_wtss_data_p1, "wtss_log2_de", "mean_alexa_log2_DE", "WTSS", "ALEXA")
test_corr(alexa_affy_data_p1, "mip_v_fur_DE_U", "mean_plier_log2_DE", "ALEXA", "AFFY")

#determine the genes common to ALL 3 platforms and limit the plots to those - 2434 genes total
common_genes_1 = which(affy_wtss_data_p1[,"ALEXA_ID"] %in% alexa_wtss_data_p1[,"ALEXA_ID"])
common_genes_2 = which(alexa_wtss_data_p1[,"ALEXA_ID"] %in% affy_wtss_data_p1[,"ALEXA_ID"])
common_genes_3 = which(alexa_affy_data_p1[,"ALEXA_ID"] %in% affy_wtss_data_p1[,"ALEXA_ID"])

length(common_genes_1) 
length(common_genes_2)
length(common_genes_3)

affy_wtss_data_common = affy_wtss_data_p1[common_genes_1,]
alexa_wtss_data_common = alexa_wtss_data_p1[common_genes_2,]
alexa_affy_data_common = alexa_affy_data_p1[common_genes_3,]

# 3-way, plus1 values only, COMMON genes only
par(mfrow=c(2,2))
test_corr(affy_wtss_data_common, "wtss_log2_de_n1", "mean_plier_log2_DE", "WTSS", "AFFY")
test_corr(alexa_wtss_data_common, "wtss_log2_de_n1", "mean_alexa_log2_DE", "WTSS", "ALEXA")
test_corr(alexa_affy_data_common, "mip_v_fur_DE_U", "mean_plier_log2_DE", "ALEXA", "AFFY")

############################################################################################################################
#Which WTSS expression measure produces the best DE correlations to array data: Read Count, RAW Coverage, Norm1, Norm2, Norm3

#Affy
par(mfrow=c(2,2))
test_corr(affy_wtss_data_common, "wtss_log2_de_rc", "mean_plier_log2_DE", "WTSS Read Count", "AFFY")
#test_corr(affy_wtss_data_common, "wtss_log2_de_raw", "mean_plier_log2_DE", "WTSS RAW Coverage", "AFFY")
test_corr(affy_wtss_data_common, "wtss_log2_de_n1", "mean_plier_log2_DE", "WTSS NORM1", "AFFY")
test_corr(affy_wtss_data_common, "wtss_log2_de_n2", "mean_plier_log2_DE", "WTSS NORM2", "AFFY")
test_corr(affy_wtss_data_common, "wtss_log2_de_n3", "mean_plier_log2_DE", "WTSS NORM3", "AFFY")


#ALEXA
par(mfrow=c(2,2))
test_corr(alexa_wtss_data_common, "wtss_log2_de_rc", "mean_alexa_log2_DE", "WTSS Read Count", "ALEXA")
#test_corr(alexa_wtss_data_common, "wtss_log2_de_raw", "mean_alexa_log2_DE", "WTSS RAW Coverage", "ALEXA")
test_corr(alexa_wtss_data_common, "wtss_log2_de_n1", "mean_alexa_log2_DE", "WTSS NORM1", "ALEXA")
test_corr(alexa_wtss_data_common, "wtss_log2_de_n2", "mean_alexa_log2_DE", "WTSS NORM2", "ALEXA")
test_corr(alexa_wtss_data_common, "wtss_log2_de_n3", "mean_alexa_log2_DE", "WTSS NORM3", "ALEXA")


#What is the effect of requiring increasing Illumina coverage on the Affy vs. WTSS comparison
test1 = affy_wtss_data_p1[which(affy_wtss_data_p1[,"MIP101_Average_Coverage_RAW"] > -1 | affy_wtss_data_p1[,"MIP5FUR_Average_Coverage_RAW"] > -1),]
test2 = affy_wtss_data_p1[which(affy_wtss_data_p1[,"MIP101_Average_Coverage_RAW"] > 0 | affy_wtss_data_p1[,"MIP5FUR_Average_Coverage_RAW"] > 0),]
test3 = affy_wtss_data_p1[which(affy_wtss_data_p1[,"MIP101_Average_Coverage_RAW"] > 10 | affy_wtss_data_p1[,"MIP5FUR_Average_Coverage_RAW"] > 10),]
test4 = affy_wtss_data_p1[which(affy_wtss_data_p1[,"MIP101_Average_Coverage_RAW"] > 100 | affy_wtss_data_p1[,"MIP5FUR_Average_Coverage_RAW"] > 100),]

par(mfrow=c(2,2))
test_corr(test1, "wtss_log2_de", "mean_plier_log2_DE", "WTSS > -1", "AFFY") #no cutoff
test_corr(test2, "wtss_log2_de", "mean_plier_log2_DE", "WTSS > 0", "AFFY")  #expression > 0
test_corr(test3, "wtss_log2_de", "mean_plier_log2_DE", "WTSS > 10", "AFFY")  #expression > 10
test_corr(test4, "wtss_log2_de", "mean_plier_log2_DE", "WTSS > 100", "AFFY")  #expression > 100

#Determine the number of DE genes (>= 2-fold) WITHIN each platform and the overlap of these genes BETWEEN each platform 
fold_change_cutoff = log2(2)

affy_de_genes = affy_wtss_data_common[which(abs(affy_wtss_data_common[,"mean_plier_log2_DE"]) >= fold_change_cutoff),]
length(affy_de_genes[,"ALEXA_ID"])

alexa_de_genes = alexa_wtss_data_common[which(abs(alexa_wtss_data_common[,"mean_alexa_log2_DE"]) >= fold_change_cutoff),] 
length(alexa_de_genes[,"ALEXA_ID"])

wtss_de_genes = affy_wtss_data_common[which(abs(affy_wtss_data_common[,"wtss_log2_de"]) >= fold_change_cutoff),] 
length(wtss_de_genes[,"ALEXA_ID"])

#Get a list of genes DE > cutoff in any platform
x = c(affy_de_genes[,"ALEXA_ID"], alexa_de_genes[,"ALEXA_ID"], wtss_de_genes[,"ALEXA_ID"]) 
de_gene_list = unique (x)
length(x)
length(unique(x))

affy_de_genes = affy_wtss_data_common[which(affy_wtss_data_common[,"ALEXA_ID"] %in% de_gene_list),]
length(affy_de_genes[,"ALEXA_ID"])

alexa_de_genes = alexa_wtss_data_common[which(alexa_wtss_data_common[,"ALEXA_ID"] %in% de_gene_list),] 
length(alexa_de_genes[,"ALEXA_ID"])

wtss_de_genes = affy_wtss_data_common[which(affy_wtss_data_common[,"ALEXA_ID"] %in% de_gene_list),] 
length(wtss_de_genes[,"ALEXA_ID"])

data = data.frame(affy_de_genes[,c("ALEXA_ID","EnsEMBL_Gene_ID","Gene_Name","mean_plier_log2_DE")], alexa_de_genes[,c("mean_alexa_log2_DE","wtss_log2_de")])
write.table (data[,], "DE_GeneList_3Platforms.txt", sep="\t", row.names = FALSE)


#Determine overlap of DE genes > cutoff between platforms - COMMON Genes only
#fold_change_cutoff = log2(4)
affy_de_genes = affy_wtss_data_common[which(abs(affy_wtss_data_common[,"mean_plier_log2_DE"]) >= fold_change_cutoff), "ALEXA_ID"]
alexa_de_genes = alexa_wtss_data_common[which(abs(alexa_wtss_data_common[,"mean_alexa_log2_DE"]) >= fold_change_cutoff), "ALEXA_ID"] 
wtss_de_genes = affy_wtss_data_common[which(abs(affy_wtss_data_common[,"wtss_log2_de_n1"]) >= fold_change_cutoff), "ALEXA_ID"] 

length(affy_de_genes)                                                                       #AFFY
length(alexa_de_genes)                                                                      #ALEXA
length(wtss_de_genes)                                                                       #WTSS
length(which(affy_de_genes %in% alexa_de_genes))                                            #AFFY and ALEXA
length(which(affy_de_genes %in% wtss_de_genes))                                             #AFFY and WTSS
length(which(wtss_de_genes %in% alexa_de_genes))                                            #WTSS and ALEXA
length(which(wtss_de_genes %in% affy_de_genes[which(affy_de_genes %in% alexa_de_genes)]))   #AFFY and ALEXA and WTSS

#Determine the overlap between the 'top 50' list of each platform
affy_de_genes = affy_wtss_data_common[(order(abs(affy_wtss_data_common[,"mean_plier_log2_DE"]), decreasing=TRUE)[1:50]), "ALEXA_ID"]
alexa_de_genes = alexa_wtss_data_common[(order(abs(alexa_wtss_data_common[,"mean_alexa_log2_DE"]), decreasing=TRUE)[1:50]) , "ALEXA_ID"]
wtss_de_genes = affy_wtss_data_common[(order(abs(affy_wtss_data_common[,"wtss_log2_de_n1"]), decreasing=TRUE)[1:50]) , "ALEXA_ID"]





############################################################################################################################
#Determine overlap of DE genes > cutoff between platforms - ALL Genes
affy_de_genes = affy_wtss_data_p1[which(abs(affy_wtss_data_p1[,"mean_plier_log2_DE"]) >= fold_change_cutoff), "ALEXA_ID"]
alexa_de_genes = alexa_wtss_data_p1[which(abs(alexa_wtss_data_p1[,"mean_alexa_log2_DE"]) >= fold_change_cutoff), "ALEXA_ID"] 
wtss_de_genes = wtss_data_all_genes_p1[which(abs(wtss_data_all_genes_p1[,"wtss_log2_de_n1"]) >= fold_change_cutoff), "ALEXA_ID"] 

length(affy_de_genes)                                                                       #AFFY
length(alexa_de_genes)                                                                      #ALEXA
length(wtss_de_genes)                                                                       #WTSS
length(which(affy_de_genes %in% alexa_de_genes))                                            #AFFY and ALEXA
length(which(affy_de_genes %in% wtss_de_genes))                                             #AFFY and WTSS
length(which(wtss_de_genes %in% alexa_de_genes))                                            #WTSS and ALEXA
length(which(wtss_de_genes %in% affy_de_genes[which(affy_de_genes %in% alexa_de_genes)]))   #AFFY and ALEXA and WTSS


#How many genes were actually detected or detectable by each platform?
length(affy_wtss_data_p1[,"ALEXA_ID"])
length(alexa_wtss_data_p1[,"ALEXA_ID"])
length(wtss_data_all_genes_p1[,"ALEXA_ID"])




############################################################################################################################
#Which WTSS expression measure produces the best Expression correlations to array data: Read Count, RAW Coverage, Norm1, Norm2
#Use MIP101 values for this comparison (library with greater depth)

#Function to create correlation plot for expression values
test_corr_exp = function(data, x_name, y_name, x_label, y_label, library){
  x = log2(data[, x_name] + 1)
  #x = data[, x_name] + 1
  y = data[, y_name]

  result1 = cor(cbind(x, y), method="pearson")
  result2 = cor(cbind(x, y), method="spearman")

  gene_count = length(x)

  result = data.frame (result1[1,2], result2[1,2], gene_count) 
  names(result) = c("pearson_all", "spearman_all", "gene_count")

  new_x_label = paste(x_label, " (", result["gene_count"], " genes)")
  main_label = paste(library, " Expression")

  colors = colorRampPalette(c("white", "blue", "#007FFF", "cyan","#7FFF7F", "yellow", "#FF7F00", "red", "#7F0000"))

  smoothScatter(x, y, xlab=new_x_label, ylab=y_label, main=main_label, colramp=colors, nbin=275, ylim=c(0,max(y)), xlim=c(0,max(x)))
  #abline(0,1, col="black", lwd=2)

  #Add some text to display Pearson and Spearman correlation values
  x_pos1 = max(x) * 0.78 
  y_pos1 = max(y) * 0.10
  y_pos2 = max(y) * 0.05

  text1 = paste("Pearson = ", round(result["pearson_all"], 2))
  text2 = paste("Spearman = ", round(result["spearman_all"], 2))

  text (x_pos1, y_pos1, text1, adj=c(0,0))
  text (x_pos1, y_pos2, text2, adj=c(0,0))

  return(result)
}

#Affy - 3-way common genes only
par(mfrow=c(2,2))
test_corr_exp(affy_wtss_data_common, "MIP101_Read_Count", "MIP101_mean_plier_log2", "WTSS Read Count", "AFFY", "MIP101")
test_corr_exp(affy_wtss_data_common, "MIP101_Average_Coverage_RAW", "MIP101_mean_plier_log2", "WTSS - AVG. COV.", "AFFY", "MIP101")
test_corr_exp(affy_wtss_data_common, "MIP101_Average_Coverage_NORM1", "MIP101_mean_plier_log2", "WTSS - NORM1", "AFFY", "MIP101")
test_corr_exp(affy_wtss_data_common, "MIP101_Average_Coverage_NORM2", "MIP101_mean_plier_log2", "WTSS - NORM2", "AFFY", "MIP101")


#ALEXA - 3-way common genes only
par(mfrow=c(2,2))
test_corr_exp(alexa_wtss_data_common, "MIP101_Read_Count", "MIP101_mean_alexa_log2", "WTSS Read Count", "ALEXA", "MIP101")
test_corr_exp(alexa_wtss_data_common, "MIP101_Average_Coverage_RAW", "MIP101_mean_alexa_log2", "WTSS - AVG. COV.", "ALEXA", "MIP101")
test_corr_exp(alexa_wtss_data_common, "MIP101_Average_Coverage_NORM1", "MIP101_mean_alexa_log2", "WTSS - NORM1", "ALEXA", "MIP101")
test_corr_exp(alexa_wtss_data_common, "MIP101_Average_Coverage_NORM2", "MIP101_mean_alexa_log2", "WTSS - NORM2", "ALEXA", "MIP101")

#Three-way comparison - 3-way common genes only
par(mfrow=c(2,2))
test_corr_exp(affy_wtss_data_common, "MIP101_Average_Coverage_RAW", "MIP101_mean_plier_log2", "WTSS", "AFFY", "MIP101")
test_corr_exp(alexa_wtss_data_common, "MIP101_Average_Coverage_RAW", "MIP101_mean_alexa_log2", "WTSS", "ALEXA", "MIP101")
test_corr_exp(alexa_affy_data_common, "MIP101_mean_plier_log2", "mip_LOG2_I_U", "AFFY", "ALEXA", "MIP101")
test_corr_exp(alexa_affy_data_common, "mip_LOG2_I_U", "MIP101_mean_plier_log2", "ALEXA", "AFFY", "MIP101")

#Affy vs WTSS (all genes profiled by the affymetrix exon arrays)
affy_wtss_data_p1[1,]
affy_wtss_data_p1_expressed = affy_wtss_data_p1[which(affy_wtss_data_p1[,"MIP101_Average_Coverage_RAW"] > 10),]
affy_wtss_data_p1_expressed_both = affy_wtss_data_p1[which(affy_wtss_data_p1[,"MIP101_Average_Coverage_RAW"] > 10 & affy_wtss_data_p1[,"MIP101_mean_plier_log2"] > 3.46),]

test_corr_exp(affy_wtss_data_p1, "MIP101_Average_Coverage_RAW", "MIP101_mean_plier_log2", "WTSS", "AFFY", "MIP101")
test_corr_exp(affy_wtss_data_p1_expressed, "MIP101_Average_Coverage_RAW", "MIP101_mean_plier_log2", "WTSS", "AFFY", "MIP101")
test_corr_exp(affy_wtss_data_p1_expressed_both, "MIP101_Average_Coverage_RAW", "MIP101_mean_plier_log2", "WTSS", "AFFY", "MIP101")








